## Feature Selection

In [19]:
import pandas as pd

### Merge Feature and Target Dataframes

In [25]:
df_features_census = pd.read_csv('./census_processed.csv')
df_features_census["geo_12"] = df_features_census["GEO_ID"].map(lambda x: str(x)[-12:])
df_features_census.drop(["GEO_ID"], axis=1, inplace=True)
df_features_census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3993 entries, 0 to 3992
Data columns (total 16 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Median_Income                            3836 non-null   float64
 1   Median_Age                               3993 non-null   float64
 2   Percent_NonCitizen                       3989 non-null   float64
 3   Percent_SpeakEngl_Poorly                 3989 non-null   float64
 4   Percent_less_than_HS                     3989 non-null   float64
 5   Percent_HS                               3989 non-null   float64
 6   Percent_SomeCollege                      3989 non-null   float64
 7   Percent_Bach                             3989 non-null   float64
 8   Percent_Grad                             3989 non-null   float64
 9   Percent_No_vehicals                      3989 non-null   float64
 10  Percent_Received_SNAP                    3989 no

In [27]:
df_features_places = pd.read_csv('./places_count_by_census_block.csv', dtype={'geo_12': 'str'})
df_features_places.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2194 entries, 0 to 2193
Data columns (total 11 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   geo_12                                               2194 non-null   object 
 1   automotive_repair_and_maintenance                    2194 non-null   float64
 2   child_day_care_services                              2194 non-null   float64
 3   elementary_and_secondary_schools                     2194 non-null   float64
 4   grocery_stores                                       2194 non-null   float64
 5   health_and_personal_care_stores                      2194 non-null   float64
 6   museums,_historical_sites,_and_similar_institutions  2194 non-null   float64
 7   offices_of_physicians                                2194 non-null   float64
 8   other_amusement_and_recreation_industries            2194 non-null  

In [28]:
df_targets = pd.read_csv('./COVID_mobility_targets.csv', dtype={'geo_12': 'str'})
df_targets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87095 entries, 0 to 87094
Data columns (total 6 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   geo_12                    87095 non-null  object 
 1   Week                      87095 non-null  int64  
 2   fraction_of_devices_home  87095 non-null  float64
 3   fraction_of_devices_work  87095 non-null  float64
 4   fraction_time_away_all    87095 non-null  float64
 5   fraction_time_away_leave  87095 non-null  float64
dtypes: float64(4), int64(1), object(1)
memory usage: 4.0+ MB


In [29]:
df_features = df_features_places.merge(df_features_census, on='geo_12')
df_features.head()

geo_12  automotive_repair_and_maintenance  child_day_care_services  \
0  170310101001                                0.0                      0.0   
1  170310101002                                0.0                      1.0   
2  170310101003                                0.0                      1.0   
3  170310102011                                0.0                      0.0   
4  170310102012                                1.0                      4.0   

   elementary_and_secondary_schools  grocery_stores  \
0                               0.0             0.0   
1                               1.0             0.0   
2                               0.0             0.0   
3                               0.0             0.0   
4                               1.0             1.0   

   health_and_personal_care_stores  \
0                              0.0   
1                              0.0   
2                              0.0   
3                              0.0   
4                              1.0   

   museums,_historical_sites,_and_similar_institutions  offices_of_physicians  \
0                                                0.0                      0.0   
1                                                3.0                      1.0   
2                                                3.0                      0.0   
3                                                0.0                      0.0   
4                                                0.0                      0.0   

   other_amusement_and_recreation_industries  religious_organizations  ...  \
0                                        0.0                      0.0  ...   
1                                        2.0                      1.0  ...   
2                                        1.0                      1.0  ...   
3                                        0.0                      0.0  ...   
4                                        0.0                      0.0  ...   

   Percent_HS  Percent_SomeCollege  Percent_Bach  Percent_Grad  \
0    0.168511             0.255138      0.289598      0.179260   
1    0.168511             0.255138      0.289598      0.179260   
2    0.168511             0.255138      0.289598      0.179260   
3    0.246110             0.246314      0.193898      0.114251   
4    0.246110             0.246314      0.193898      0.114251   

   Percent_No_vehicals  Percent_Received_SNAP  \
0             0.344589               0.186960   
1             0.344589               0.186960   
2             0.344589               0.186960   
3             0.140014               0.316592   
4             0.140014               0.316592   

   Percent_Men_Usually_Fulltime_Employed  \
0                               0.885906   
1                               0.507064   
2                               0.765318   
3                               0.559184   
4                               0.495495   

   Percent_Women_Usually_Fulltime_Employed  Percent_No_Internet_Access  \
0                                 0.382653                    0.217742   
1                                 0.452071                    0.330517   
2                                 0.663338                    0.074041   
3                                 0.437107                    0.105030   
4                                 0.320800                    0.202247   

   Percent_Computing_Device  
0                  0.782258  
1                  0.829989  
2                  0.950045  
3                  0.921598  
4                  0.822868  

[5 rows x 26 columns]

In [33]:
df = df_features.merge(df_targets, on='geo_12')
df.columns

Index(['geo_12', 'automotive_repair_and_maintenance',
       'child_day_care_services', 'elementary_and_secondary_schools',
       'grocery_stores', 'health_and_personal_care_stores',
       'museums,_historical_sites,_and_similar_institutions',
       'offices_of_physicians', 'other_amusement_and_recreation_industries',
       'religious_organizations', 'restaurants_and_other_eating_places',
       'Median_Income', 'Median_Age', 'Percent_NonCitizen',
       'Percent_SpeakEngl_Poorly', 'Percent_less_than_HS', 'Percent_HS',
       'Percent_SomeCollege', 'Percent_Bach', 'Percent_Grad',
       'Percent_No_vehicals', 'Percent_Received_SNAP',
       'Percent_Men_Usually_Fulltime_Employed',
       'Percent_Women_Usually_Fulltime_Employed', 'Percent_No_Internet_Access',
       'Percent_Computing_Device', 'Week', 'fraction_of_devices_home',
       'fraction_of_devices_work', 'fraction_time_away_all',
       'fraction_time_away_leave'],
      dtype='object')

#### For feature selection, we can just drop rows with nas.

In [43]:
df.dropna(inplace=True)

### Look at correlations between features and targets to get an idea about which features might be important

In [44]:
df.corr()

automotive_repair_and_maintenance  \
automotive_repair_and_maintenance                                        1.000000e+00   
child_day_care_services                                                  8.508338e-02   
elementary_and_secondary_schools                                         4.178819e-02   
grocery_stores                                                           1.114921e-01   
health_and_personal_care_stores                                          7.421685e-02   
museums,_historical_sites,_and_similar_institut...                       3.194654e-02   
offices_of_physicians                                                    1.788637e-02   
other_amusement_and_recreation_industries                                8.078375e-02   
religious_organizations                                                 -1.812855e-03   
restaurants_and_other_eating_places                                      5.433176e-02   
Median_Income                                                           -2.184869e-02   
Median_Age                                                              -5.062362e-02   
Percent_NonCitizen                                                       1.196121e-01   
Percent_SpeakEngl_Poorly                                                 1.604906e-01   
Percent_less_than_HS                                                     1.469316e-01   
Percent_HS                                                               8.583942e-02   
Percent_SomeCollege                                                     -4.064785e-02   
Percent_Bach                                                            -7.550529e-02   
Percent_Grad                                                            -1.021233e-01   
Percent_No_vehicals                                                     -1.659982e-01   
Percent_Received_SNAP                                                   -2.070637e-02   
Percent_Men_Usually_Fulltime_Employed                                    4.338496e-02   
Percent_Women_Usually_Fulltime_Employed                                 -1.811937e-02   
Percent_No_Internet_Access                                               2.233448e-02   
Percent_Computing_Device                                                -1.752393e-02   
Week                                                                    -6.451187e-18   
fraction_of_devices_home                                                 2.472376e-02   
fraction_of_devices_work                                                -7.427150e-02   
fraction_time_away_all                                                  -5.077652e-02   
fraction_time_away_leave                                                -5.175712e-02   

                                                    child_day_care_services  \
automotive_repair_and_maintenance                              8.508338e-02   
child_day_care_services                                        1.000000e+00   
elementary_and_secondary_schools                               1.656755e-01   
grocery_stores                                                 1.218192e-01   
health_and_personal_care_stores                                1.550656e-01   
museums,_historical_sites,_and_similar_institut...             2.091628e-01   
offices_of_physicians                                          4.963597e-02   
other_amusement_and_recreation_industries                      1.446378e-01   
religious_organizations                                        1.491348e-01   
restaurants_and_other_eating_places                            1.941497e-01   
Median_Income                                                 -1.970866e-02   
Median_Age                                                     9.529237e-04   
Percent_NonCitizen                                            -5.516027e-02   
Percent_SpeakEngl_Poorly                                      -7.010756e-02   
Percent_less_than_HS                                          -3.302697e-02   
Percent_HS                  

#### Looks like a lot of correlations that are not very strong.

In [45]:
df.describe().loc[['mean', 'std']]

automotive_repair_and_maintenance  child_day_care_services  \
mean                           0.391534                 0.360750   
std                            0.874902                 0.680763   

      elementary_and_secondary_schools  grocery_stores  \
mean                          0.432900        0.471861   
std                           0.692307        0.980868   

      health_and_personal_care_stores  \
mean                         0.365079   
std                          0.998801   

      museums,_historical_sites,_and_similar_institutions  \
mean                                           0.306878     
std                                            0.881180     

      offices_of_physicians  other_amusement_and_recreation_industries  \
mean               0.582492                                   0.583935   
std                2.977112                                   1.685666   

      religious_organizations  restaurants_and_other_eating_places  ...  \
mean                 0.658971                             3.432900  ...   
std                  1.031800                             8.506581  ...   

      Percent_Received_SNAP  Percent_Men_Usually_Fulltime_Employed  \
mean               0.213511                               0.630314   
std                0.161835                               0.195814   

      Percent_Women_Usually_Fulltime_Employed  Percent_No_Internet_Access  \
mean                                 0.517987                    0.204684   
std                                  0.169586                    0.139384   

      Percent_Computing_Device       Week  fraction_of_devices_home  \
mean                  0.845526  16.000000                  0.461096   
std                   0.117806   2.582058                  0.086326   

      fraction_of_devices_work  fraction_time_away_all  \
mean                  0.080644                0.139449   
std                   0.032823                0.050988   

      fraction_time_away_leave  
mean                  0.259828  
std                   0.071834  

[2 rows x 30 columns]

In [152]:
from sklearn.model_selection import train_test_split

features = ['automotive_repair_and_maintenance',
            'child_day_care_services', 'elementary_and_secondary_schools',
            'grocery_stores', 'health_and_personal_care_stores',
            'museums,_historical_sites,_and_similar_institutions',
            'offices_of_physicians', 'other_amusement_and_recreation_industries',
            'religious_organizations', 'restaurants_and_other_eating_places',
            'Median_Income', 'Median_Age', 'Percent_NonCitizen',
            'Percent_SpeakEngl_Poorly', 'Percent_less_than_HS', 'Percent_HS',
            'Percent_SomeCollege', 'Percent_Bach', 'Percent_Grad',
            'Percent_No_vehicals', 'Percent_Received_SNAP',
            'Percent_Men_Usually_Fulltime_Employed',
            'Percent_Women_Usually_Fulltime_Employed', 'Percent_No_Internet_Access',
            'Percent_Computing_Device']
target = 'fraction_time_away_all'

X = df[features].values
y = df[target].values

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2)

### Linear Regression

In [153]:
#Normalize
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [154]:
from sklearn import linear_model
regr = linear_model.LinearRegression()

regr.fit(X_train_scaled, y_train)

LinearRegression()

In [155]:
y_pred = regr.predict(X_test_scaled)

In [156]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_pred, y_test)

0.0022030595351213403

In [157]:
from sklearn.metrics import r2_score

regr.score(X_train_scaled, y_train)

0.15565658274362693

In [158]:
pd.Series(features)[list(abs(regr.coef_) > 0.001)]

2            elementary_and_secondary_schools
3                              grocery_stores
6                       offices_of_physicians
9         restaurants_and_other_eating_places
10                              Median_Income
11                                 Median_Age
12                         Percent_NonCitizen
13                   Percent_SpeakEngl_Poorly
14                       Percent_less_than_HS
16                        Percent_SomeCollege
17                               Percent_Bach
18                               Percent_Grad
19                        Percent_No_vehicals
20                      Percent_Received_SNAP
21      Percent_Men_Usually_Fulltime_Employed
22    Percent_Women_Usually_Fulltime_Employed
dtype: object

In [159]:
feature_importances_regr = pd.DataFrame(regr.coef_,
                                   index = df[features].columns,
                                    columns=['importance']).sort_values('importance', ascending=False)

#regr.coef_
feature_importances_regr

importance
Percent_NonCitizen                                    0.008536
Percent_No_vehicals                                   0.007544
Percent_Grad                                          0.005820
restaurants_and_other_eating_places                   0.004979
Median_Income                                         0.002161
offices_of_physicians                                 0.002006
Percent_less_than_HS                                  0.001936
elementary_and_secondary_schools                      0.001215
child_day_care_services                               0.000623
other_amusement_and_recreation_industries             0.000592
museums,_historical_sites,_and_similar_institut...    0.000365
religious_organizations                               0.000250
Percent_Computing_Device                              0.000063
Percent_No_Internet_Access                           -0.000227
automotive_repair_and_maintenance                    -0.000444
Percent_HS                                           -0.000474
health_and_personal_care_stores                      -0.000974
grocery_stores                                       -0.001344
Percent_Bach                                         -0.003378
Percent_SomeCollege                                  -0.004304
Median_Age                                           -0.005621
Percent_Men_Usually_Fulltime_Employed                -0.006261
Percent_Women_Usually_Fulltime_Employed              -0.006540
Percent_Received_SNAP                                -0.007669
Percent_SpeakEngl_Poorly                             -0.015919

For target: fraction_time_away_leave
Percent_Grad 	0.012660
Percent_No_vehicals 	0.011007
Percent_NonCitizen 	0.008608
restaurants_and_other_eating_places 	0.003511
other_amusement_and_recreation_industries 	0.002147
child_day_care_services 	0.001538
offices_of_physicians 	0.001052
grocery_stores 	-0.002207
Percent_less_than_HS 	-0.002730
Percent_HS 	-0.003446
Percent_Received_SNAP 	-0.004576
Median_Age 	-0.005178
Percent_Men_Usually_Fulltime_Employed 	-0.006382
Percent_Women_Usually_Fulltime_Employed 	-0.006666
Percent_SomeCollege 	-0.007579
Percent_SpeakEngl_Poorly 	-0.008516


For target: fraction_of_devices_home
Percent_SpeakEngl_Poorly 	0.042864
Percent_Bach 	0.012126
Percent_Received_SNAP 	0.011084
Median_Age 	0.008225
Percent_Women_Usually_Fulltime_Employed 	0.004302
Percent_Grad 	0.004125
child_day_care_services 	0.002590
Percent_Men_Usually_Fulltime_Employed 	0.002115
grocery_stores 	-0.001893
Percent_No_Internet_Access 	-0.002190
Median_Income 	-0.002456
elementary_and_secondary_schools 	-0.003602
Percent_No_vehicals 	-0.004472
offices_of_physicians 	-0.005166
restaurants_and_other_eating_places 	-0.006488
Percent_HS 	-0.006724
Percent_NonCitizen 	-0.012125
Percent_less_than_HS 	-0.013243

For target: fraction_of_devices_work
Percent_less_than_HS 	0.004021
Percent_No_vehicals 	0.003656
Median_Income 	0.002395
Median_Age 	0.001979
Percent_Men_Usually_Fulltime_Employed 	0.001618
automotive_repair_and_maintenance 	-0.001159
religious_organizations 	-0.001167
child_day_care_services 	-0.001319
Percent_Bach 	-0.001738
other_amusement_and_recreation_industries 	-0.002410
grocery_stores 	-0.002565
Percent_SomeCollege 	-0.003431
Percent_SpeakEngl_Poorly 	-0.005126

For target: fraction_time_away_all
Percent_NonCitizen 	0.008536
Percent_No_vehicals 	0.007544
Percent_Grad 	0.005820
restaurants_and_other_eating_places 	0.004979
Median_Income 	0.002161
offices_of_physicians 	0.002006
Percent_less_than_HS 	0.001936
elementary_and_secondary_schools 	0.001215
grocery_stores 	-0.001344
Percent_Bach 	-0.003378
Percent_SomeCollege 	-0.004304
Median_Age 	-0.005621
Percent_Men_Usually_Fulltime_Employed 	-0.006261
Percent_Women_Usually_Fulltime_Employed 	-0.006540
Percent_Received_SNAP 	-0.007669
Percent_SpeakEngl_Poorly 	-0.015919

### Random Forest Model

In [131]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_jobs=-1, 
                            n_estimators=100,
                           min_samples_leaf=1
                           )
model = rf.fit(X_train,y_train)

In [132]:
y_pred=rf.predict(X_test)

In [133]:
mean_squared_error(y_pred, y_test)

0.0030394838746226616

This is lower than the MSE for the linear regression model.

In [124]:
#Show how much each feature contributes to the model prediction
feature_importances = pd.DataFrame(rf.feature_importances_,
                                   index = df[features].columns,
                                    columns=['importance']).sort_values('importance', ascending=False)
print(feature_importances)

                                                    importance
Percent_Grad                                          0.202535
Median_Age                                            0.161760
Percent_HS                                            0.069688
Percent_No_vehicals                                   0.067948
Percent_Men_Usually_Fulltime_Employed                 0.055196
Median_Income                                         0.054337
Percent_Women_Usually_Fulltime_Employed               0.051006
Percent_No_Internet_Access                            0.042352
Percent_Computing_Device                              0.039522
Percent_Bach                                          0.033824
Percent_SomeCollege                                   0.033617
Percent_Received_SNAP                                 0.025764
Percent_NonCitizen                                    0.024985
Percent_less_than_HS                                  0.022956
Percent_SpeakEngl_Poorly                              0